link-grammar/connectors.h

```diff
@@ -97,11 +100,24 @@ struct condesc_struct
 	/* For connector match speedup when sorting the connector table. */
 	uint8_t uc_length;   /* uc part length */
 	uint8_t uc_start;    /* uc start position */
-};
-typedef struct condesc_struct condesc_t;
+} hdesc_t;
 
-/* Length-limits for how far connectors can reach out. */
-#define UNLIMITED_LEN 255
+/* Each connector type has a connector descriptor. The size of this
+ * struct is 32 byes, to facilitate CPU memory caching during parsing.
+ * The "more" field points to connector information that is needed in
+ * other steps. The con_num field is used a lot in steps that need
+ * connector hashing, and it is included here to avoid extra
+ * redirections.
+ * Multi connectors are considering the same type as their non-multi
+ * version, so the multi indication is kept in Connector_struct. */
+struct condesc_struct
+{
+	lc_enc_t lc_letters;
+	lc_enc_t lc_mask;
+	hdesc_t *more;       /* More information, for keeping small struct size. */
```

Comment:
So `hdesc` points at `condesc`, and `condesc` points at `hdesc`. Interesting ... the old code did not have this loop.

Reply:
`hdesc` are the slots of the connector hash table. Each one that is used, has a `desc` pointer to a `condesc_t` element. Up to here it is a standard implementation of an open hash table.
The `more` elements of `condesc_t` need to point somewhere, and it was convenient to use the `hdesc` slots, which are already allocated (it requires less code). However, this is memory inefficient (and indirectly CPU inefficiently due to cache trashing) because most of the `hdesc` elements are empty (NULL `desc`).

The solution is to use a memory pool for the `more` elements. I will eventually send a cleanup PR for that (and the  `sort_condesc_by_uc_constring()` code.

Reply:
Oh! Ah! OK. I think I understand.  This wasn't obvious from casual perusal, so I thought I'd point it out.